### DISCUSSION TITLE
[Script request] NixOS VM & LXC

### DISCUSSION THREAD

**Discussion by alexanderjerome** [community-scripts/ProxmoxVE#764] [Request script]
### Application Name

NixOS VM/LXC

### Website

https://nixos.org/

### Description

A NixOS automation script for Proxmox VE would bridge the gap between declarative infrastructure and ease of use, making reproducible homelab environments accessible to a broader user base. By automating the complex installation process, users can focus on defining their desired system state rather than wrestling with technical setup procedures. Moreover, integrating NixOS into Proxmox Helper-Scripts could revolutionize the project's package maintenance approach - leveraging the vast nixpkgs ecosystem and flakes for OS-agnostic package management would not only attract the growing Nix community but could also streamline the maintenance of helper scripts themselves, creating a more robust and maintainable platform for all users.

### Due Diligence

- [X] I have searched existing [scripts](https://community-scripts.github.io/Proxmox/scripts) and found no duplicates.
- [X] I have searched existing [discussions](https://github.com/community-scripts/ProxmoxVE/discussions?discussions_q=) and found no duplicate requests.
- [X] This is not a game-related request.

**Comment by MickLesk** [comment]
Adding an LXC is quite pointless. It is an oneliner that you have to add to Debian Ubuntu LXC. VM could be considered, but I don't see any added value for >90% of helper script users yet

**Reply by AskAlice** [reply]
I think having one source of truth for adding well-known OS's, images, containers cna be useful, but this might be only less useful than installing netboot.xyz. i suppose in terms of nix they can be always kept up to date with rolling releases as the whole OS is declarative and built for gitops.

personally athena is more useful in general 
https://athenaos.org/en/installation/install-nixos/

**Reply by alexanderjerome** [reply]
I'd still like to make a case for including both VM and LXC support. The current Debian LXC approach in the community scripts, while functional, requires users to maintain snapshots or templates to preserve configurations. In contrast, a NixOS LXC has less overhead and can support GitOps workflows as @AskAlice mentioned.

I haven't used Athena, but it seems like it may be Arch using [nixos-anywhere](https://github.com/nix-community/nixos-anywhere)?

The latest NixOS [LXC](https://hydra.nixos.org/job/nixos/release-24.11/nixos.proxmoxLXC.x86_64-linux/latest) and [VM](https://hydra.nixos.org/job/nixos/release-24.11/nixos.proxmoxImage.x86_64-linux/latest) Proxmox builds from Hydra are very fast to deploy. I have a work in progress [deployment utility on my Gitea repo](https://git.jeirslab.xyz/jeirmeister/NixOS-PVE-Deployment-Util) that demonstrates these additional steps. It configures the .conf files for the VM/LXC to ensure terminal access properly loads in the Proxmox UI, and temporarily mounts a nix-configuration directory into the container for the initial rebuild process. The configuration in the repo provides basic required settings for ensuring Proxmox compatibility by suppressing certain system services and also enables VS Code-Server workarounds. Eventually, I want to enable the utility to automate installation via fetching flake-based configurations from repositories, and implement an Proxmoxer API wrapper so that I can automate NixOS deployment from my dev machine for DevOps scenarios.

I'm happy to submit a PR for a community-scripts version of this in the near future, I just would need to find documentation on proper implementation into the community-scripts script templates and required build functions.

**Reply by s-celles** [reply]
No sure if there is no value about adding it.... Because I was looking for the exact same features
https://mtlynch.io/notes/nixos-proxmox/ provides a way to install NixOS using LXC (I haven't test it currently)
